### PR TITLE
fix resuming from checkpoint for fault-tolerant in case of no failure

### LINKED ISSUE #8835: resume_from_checkpoint broken when fault-tolerant feature enabled
## 🐛 Bug

A trainer with the resume from checkpoint option does not continue training and stops immediately, despite increased max_epoch settings.

### To Reproduce
```python

import os
import unittest.mock

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        print(batch.sum())
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def on_train_epoch_end(self):
        print("epoch ended")

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

    def on_load_checkpoint(self, checkpoint):
        pass


@unittest.mock.patch.dict(os.environ, {"PL_FAULT_TOLERANT_TRAINING": "1"})
def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=3,
        limit_val_batches=0,
        num_sanity_val_steps=0,
        max_epochs=1,
        weights_summary=None,
    )
    trainer.fit(model, train_dataloader=train_data)

    trainer.save_checkpoint("lightning_logs/auto.pt")

    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=3,
        limit_val_batches=0,
        num_sanity_val_steps=0,
        max_epochs=3,
        weights_summary=None,
        resume_from_checkpoint="lightning_logs/auto.pt",
    )
    trainer.fit(model, train_dataloader=train_data)


if __name__ == "__main__":
    run()


```

Output: 

```python
Epoch 0:   0%|          | 0/3 [00:00<?, ?it/s] tensor(6.2218)
Epoch 0:  33%|███▎      | 1/3 [00:00<00:00, 83.40it/s, loss=-1.23, v_num=82]tensor(-9.5560)
Epoch 0:  67%|██████▋   | 2/3 [00:00<00:00, 120.73it/s, loss=-1.71, v_num=82]tensor(14.5846)
Epoch 0: 100%|██████████| 3/3 [00:00<00:00, 140.56it/s, loss=-0.136, v_num=82]epoch ended
Epoch 0: 100%|██████████| 3/3 [00:00<00:00, 98.17it/s, loss=-0.136, v_num=82] 

Restoring states from the checkpoint path at lightning_logs/auto.pt
Epoch 1: 100%|██████████| 3/3 [00:00<00:00, 22712.84it/s]epoch ended
Epoch 2: 100%|██████████| 3/3 [00:00<00:00, 20004.63it/s]epoch ended
Epoch 2: 100%|██████████| 3/3 [00:00<00:00, 3038.62it/s] 
```
Here the training_step does not get invoked on the second fit call. Instead, the epochs end immediately.

ONLY happens when PL_FAULT_TOLERANT_TRAINING=1. This is an experimental feature which is off by default.

### Expected behavior

The training continues due to the increased max_epochs setting.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #8835

When saving the loop state dict on a failure (breaking in the middle of an epoch), we want to save the progress counters exactly as they are in order to resume on the failed batch. The same applies to runs that finish successfully and with the full epoch completed. If we resume training in such a case, we don't want to resume on the last batch that we saved, but instead on the next one which is the batch 0 from the next epoch. So the counters need to reset every start of an epoch if:

- we are not restarting (regular case)
- if we are restarting, the current counters are at the limit of the loop AND we did not fail

**Note:** 
- This fix applies to master branch only (fault-tolerant training).
- Test fails on master

TODO: 
- [x] test

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: tests/loops/test_loops.py [hunk 1]
```diff
@@ -513,3 +513,129 @@ def configure_optimizers_multiple(self):
...
+    trainer.fit(model)
+
+    # number of batches that will call `optimizer.step()`
+    nbe_total_opt_steps = n_epochs * n_batches * n_optimizers
+    nbe_total_zero_grad = n_epochs * n_batches * n_optimizers
+
+    nbe_sch_steps = n_epochs
+    if n_optimizers > 1:
+        # assumes that the scheduler config is unchanged
+        # `* 1` because there is only one step-level scheduler
```
Comment (carmocca) [inline tests/loops/test_loops.py:567]: There's no `* 1`

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 2]
```diff
@@ -232,6 +232,11 @@ def on_run_end(self) -> None:
         if self._num_training_batches_reached(self.is_last_batch):
             self.update_lr_schedulers("epoch", update_plateau_schedulers=True)
 
+        # on a complete run, reset current progress counters in case we want to resume
+        self.batch_progress.current.reset()
+        self.scheduler_progress.current.reset()
+        self.batch_loop.optimizer_loop.optim_progress.reset_on_epoch()
+
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:239]: I'm not sure about this. Wouldn't this break subsequent `trainer.fit` calls where the first one didn't finish?

why can't we put it in `def reset`?

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:239]: sorry, I need to take this PR back and think about this more, mainly for other reasons but yes you have also a good point. My intention is to fix the problem where the checkpoint contains progress values that would make the loop restart on the last batch of the previous epoch. 

We are missing the logic in progress tracking to resume on the next batch, when so far we have only focused on resuming on the current batch if it failed.

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 3]
```diff
@@ -65,8 +65,7 @@ def connect(self, **kwargs: "Loop") -> None:
         raise NotImplementedError(f"{self.__class__.__name__} does not connect any child loops.")
 
     def reset(self) -> None:
-        if not self.restarting:
-            self.optim_progress.optimizer_idx = 0
+        self.optim_progress.optimizer_idx = 0
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:68]: reviewer: we don't actually have enough information to resume an optimizer loop here. so let's not pretend we can.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 4]
```diff
@@ -97,7 +97,10 @@ def reset(self) -> None:
         # track epoch output
         self._epoch_output = [[] for _ in range(self.batch_loop.num_active_optimizers(self.total_batch_idx))]
 
-        if not self.restarting:
+        failed = self.batch_progress.current.ready != self.batch_progress.current.completed
```
Comment (tchaton) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:100]: Won't this be affected by `reset_on_restart` applied while restarting there: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/base.py#L218

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:68]: the knowledge whether the idx needs to be reset to 0 or not. this depends on whether there was a failure or not. 
plus hidden state and all the output handling and also which optimizer currently is active

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:100]: I didn't know this exists, why is it there in the first place? if progress magically resets while we are loading the state dict then we are not actually loading what we intend to.

Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:100]: This was put in place to make sure progress is properly reset for current.

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:100]: But we don't want to restart in case of fault-tolerant training, unless we are actually past the last batch in the epoch. This information is not available at the time of loading the loop state dict

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py]: keeping it, but have to add done() check.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py]: removed the redundant ready != completed check

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 5]
```diff
@@ -65,7 +65,7 @@ def connect(self, **kwargs: "Loop") -> None:
         raise NotImplementedError(f"{self.__class__.__name__} does not connect any child loops.")
 
     def reset(self) -> None:
-        if not self.restarting:
+        if not self.restarting or self.done:
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:68]: on a larger scope, I will work on a test for restarting in a failed optimizer loop in a separate PR. It's on the todo list #9130

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:68]: Do you think that everywhere we check `not self.restarting` we should also add `or self.done`?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:68]: only in places where done() is determined by the current state. This is not the case for the training_epoch_loop where we also check max_steps, which is checked against global counters. That is why I specifically pulled out the check for `num_training_batches_reached`.

File: pytorch_lightning/loops/fit_loop.py [hunk 6]
```diff
@@ -263,7 +263,7 @@ def teardown(self) -> None:
 
     def on_save_checkpoint(self) -> Dict:
         state_dict = super().on_save_checkpoint()
-        # FIXME(@tchaton) Should pass has_completed=True when iterator is exhausted ?
+        # TODO: update has_completed to its proper value
```
Comment (justusschock) [inline pytorch_lightning/loops/fit_loop.py:266]: This is not related to this PR, right?

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 7]
```diff
@@ -98,7 +98,8 @@ def reset(self) -> None:
         # track epoch output
         self._epoch_output = [[] for _ in range(self.batch_loop.num_active_optimizers(self.total_batch_idx))]
 
-        if not self.restarting:
+        ended = self._num_training_batches_reached()
+        if not self.restarting or ended:
```
Comment (justusschock) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:102]:
```suggestion
        if not self.restarting or self._num_training_batches_reached():
```

nit: no need for a separate variable if only used once :)

**Review by justusschock — APPROVED** [review]
LGTM!

File: pytorch_lightning/loops/fit_loop.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/loops/fit_loop.py:266]: nope, but I will take away the responsibility from Thomas :)
